src/DynamoDBMessageRepository/DynamoDBMessageRepository.php

```diff
@@ -0,0 +1,140 @@
+
+use Aws\DynamoDb\DynamoDbClient;
+use Aws\DynamoDb\Marshaler;
+use Aws\ResultPaginator;
+use EventSauce\EventSourcing\AggregateRootId;
+use EventSauce\EventSourcing\Message;
+use EventSauce\EventSourcing\MessageRepository;
+use EventSauce\EventSourcing\Serialization\MessageSerializer;
+use EventSauce\EventSourcing\Header;
+use EventSauce\EventSourcing\UnableToPersistMessages;
+use EventSauce\EventSourcing\UnableToRetrieveMessages;
+use Ramsey\Uuid\Uuid;
+use Generator;
+
+use Throwable;
+
+class DynamoDBMessageRepository implements MessageRepository
+{
+    protected DynamoDbClient $client;
+    protected MessageSerializer $serializer;
+    protected string $tableName;
+
+    public function __construct(DynamoDbClient $client, MessageSerializer $serializer, string $tableName)
+    {
+        $this->client = $client;
+        $this->serializer = $serializer;
+        $this->tableName = $tableName;
+    }
+
+    public function persist(Message ...$messages): void
+    {
+        if (count($messages) === 0) {
+            return;
+        }
+
+        $items = [];
+        $marshaler = new Marshaler();
+
+        foreach ($messages as $message) {
+            $payload = $this->serializer->serializeMessage($message);
+            $event = $payload['headers'][Header::EVENT_ID] = $payload['headers'][Header::EVENT_ID] ?? Uuid::uuid4()->toString();
+            $aggregateRootId = $payload['headers'][Header::AGGREGATE_ROOT_ID] ?? null;
+            $eventType = $payload['headers'][Header::EVENT_TYPE] ?? null;
+            $aggregateRootVersion = $payload['headers'][Header::AGGREGATE_ROOT_VERSION] ?? 0;
+            $timeOfRecording = $payload['headers'][Header::TIME_OF_RECORDING];
+
+            $item = [
+                'event' => $marshaler->marshalValue($event),
+                'aggregateRootId' => $marshaler->marshalValue($aggregateRootId),
+                'eventType' => $marshaler->marshalValue($eventType),
+                'aggregateRootVersion' => $marshaler->marshalValue($aggregateRootVersion),
+                'timeOfRecording' => $marshaler->marshalValue($timeOfRecording),
+                'payload' => $marshaler->marshalItem(['payload' => $payload])['payload']
+            ];
+
+            $items[] = [
+                'PutRequest' => [
+                    'Item' => $item
+                ]
+            ];
+        }
+
+        $batchRequest = ['RequestItems' => [ $this->tableName => $items ]];
+
+        try {
+            $this->client->batchWriteItem($batchRequest);
+        } catch (Throwable $exception) {
+            throw UnableToPersistMessages::dueTo('', $exception);
+        }
+
+    }
+
+    public function retrieveAll(AggregateRootId $id): Generator
+    {
+        $query = [
+            'TableName' => $this->tableName,
+            'KeyConditionExpression' => 'aggregateRootId = :v1',
+            'ExpressionAttributeValues' => [
+                ':v1' => ['S' => $id->toString()],
+            ]
+        ];
```

Comment:
By default reads are eventually consistent, and I think in this case you probably want to opt-into immediate/strong consistency. It will incur a cost penalty.

```suggestion
        $query = [
            'TableName' => $this->tableName,
            'KeyConditionExpression' => 'aggregateRootId = :v1',
            'ConsistentRead' => true,
            'ExpressionAttributeValues' => [
                ':v1' => ['S' => $id->toString()],
            ]
        ];
```

Reply:
I understand what you mean. IMO `ConsistentRead` comes with a cost ([not only price but also higher latency](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html)) which might not be necessary for most of the use cases.
But If we consider that it's important to keep `ConsistentRead`, I would recommend it doing it as an optional parameter. What do you think?

Reply:
I'm not familiar enough with EventSauce to give a really tailored answer, but without the parameter you're risking missing some of the events that have recently been persisted. These might not even necessarily be the latest events (you could get v1, v2, v3, and then v5 for example, missing v4). In every ES implementation I've ever written, I've always prioritized consistency over latency/performance, because consistency issue basically jeopardize everything that happens downstream of the event/message store.

I think you'd be best off with @frankdejonge's $0.02 here, to determine if the tradeoff makes sense.

Reply:
From what I read in the docs consistent reads would be preferred by default. If you're trying to guard against domain race conditions, the eventual consistent nature of the read might give it a too long delay. If anything, if people prefer to opt into these race conditions (maybe) happening, it should be an opt-in config setting.